### PR TITLE
Seed dithering

### LINKED ISSUE #351: Feature request: allow setting the FPACK dither seed
Hi,

I'm trying to make it so my code writes bit-identical outputs every time it runs, but I'm using the compression schemes that do the subtractive dither thing, so there is a seed chosen at run-time.  It looks like CFITSIO has routines to set that seed -
`fits_set_quantize_dither()` or `fits_set_dither_seed()` etc
https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node41.html

It looks like one place to add this feature would be here,
https://github.com/esheldon/fitsio/blob/c8f8d608efbbc304ff58c9e005a1f382b61abfef/fitsio/fitsio_pywrap.c#L1376
ie similar to how the quantization level can be set,
https://github.com/esheldon/fitsio/blob/c8f8d608efbbc304ff58c9e005a1f382b61abfef/fitsio/fitsio_pywrap.c#L1483

thanks!
-dustin

### PR DESCRIPTION
**Pull request by esheldon**
closes #351

Add ability to seed the subtractive dither

### CODE REVIEW

**Comment by esheldon** [comment]
@dstndstn does this work for you?

**Review by beckermr — CHANGES_REQUESTED** [review]
Are there other parts of the lossy compression seed setting in the fits standard / cfitsio that we should be aware of to make sure our APIs here can support? We need not implement them OFC.

**Comment by esheldon** [comment]
Possibly.  This code I added only sets the seed when you create the HDU, which seems safe.  What other operations should be supported?

**Comment by beckermr** [comment]
The other options allowed by the standard are

```
-qt <level> - compute the seed from the checksum of the first tile of image pixels
-qN <level> where N is an integer between 1 and 10000 – use seed N (e.g., -q3010 4).
```

This all gets coupled into the FITS header via

```
FZDTHRSD - 'CLOCK', 'CHECKSUM', 1 - 10000
```

and here is the doc string for `fits_set_dither_seed`

```
This routine specifies the value of the offset that should be applied when calculating the random dithering when quantizing floating point iamges. A random offset should be applied to each image to avoid quantization effects when taking the difference of 2 images, or co-adding a set of images. Without this random offset, the corresponding pixel in every image will have exactly the same dithering.

offset = 0 means use the default random dithering based on system time offset = negative means randomly chose dithering based on 1st tile checksum offset = [1 - 10000] means use that particular dithering pattern

```

The default according to the fpackguide is 

```
By default, the initial seed value is randomly chosen based on the
current system clock time.
```

So we could support the following options for `dither_seed` at the python level:

- 0 or 'CLOCK' or 'clock': set seed to 0
- negative number or 'CHECKSUM' or 'checksum': set seed to -1
- 1-10000: send seed through
- error for anything else

**Comment by esheldon** [comment]
yikes, I should have red the docs first

File: fitsio/fitslib.py [hunk 1]
```diff
@@ -1800,6 +1845,26 @@ def get_qmethod(qmethod):
     return _qmethod_map[qmethod]
 
 
+def get_dither_seed(dither_seed):
+    if dither_seed is None or dither_seed in ['clock', 'CLOCK']:
```
Comment (beckermr) [inline fitsio/fitslib.py:1849]:
```suggestion
    if dither_seed is None or (isinstance(dither_seed, str) and dither_seed.lower() == 'clock'):
```

File: fitsio/fitslib.py [hunk 2]
```diff
@@ -1800,6 +1845,26 @@ def get_qmethod(qmethod):
...
+        # 0 means do not set the seed, use the system clock. This is the
+        # default
+        seed_out = 0
+    elif dither_seed in ['checksum', 'CHECKSUM']:
```
Comment (beckermr) [inline fitsio/fitslib.py:1853]:
```suggestion
    elif isinstance(dither_seed, str) and dither_seed.lower() == 'checksum':
```

File: fitsio/fitslib.py [hunk 3]
```diff
@@ -1800,6 +1845,26 @@ def get_qmethod(qmethod):
...
+        seed_out = -1
+    else:
+        # must fit in an int
+        seed_out = numpy.int32(dither_seed)
+
+        if seed_out < -1 or seed_out > 10_000:
```
Comment (beckermr) [inline fitsio/fitslib.py:1859]:
```suggestion
        if seed_out > 10_000:
```

Any negative value is allowed by the fits routine. Fun!

File: fitsio/fitslib.py [hunk 4]
```diff
@@ -1800,6 +1845,26 @@ def get_qmethod(qmethod):
...
+            raise ValueError(
+                f'Got dither_seed {seed_out}, expected -1, 0 or '
+                'a seed in range [1, 10000]'
```
Comment (beckermr) [inline fitsio/fitslib.py:1862]:
```suggestion
                f'Got dither_seed {seed_out}, expected a value < 10000'
```

**Review by beckermr — CHANGES_REQUESTED** [review]
A few tweaks but looking good. We'll need to adjust the tests to cover more cases and not error on -3. We should also test that -3 and -1 give the same answer.

File: fitsio/fitslib.py (hunk 1 above)
Comment (esheldon) [reply fitsio/fitslib.py:1849]: These are equivalent, and I actually think the first reads better

Comment (esheldon) [reply fitsio/fitslib.py:1849]: actually no, I guess you are suggesting we accept things like `CloCk`

Comment (beckermr) [reply fitsio/fitslib.py:1849]: I agree it is an odd edge case, but I can imagine someone putting in a single capitol by accident and getting super frustrated why things are broken / not working.

**Review by beckermr — APPROVED** [review]
LGTM! Thanks!

File: fitsio/fitslib.py [hunk 5]
```diff
@@ -377,6 +378,16 @@ def write(filename, data, extname=None, extver=None, header=None,
 
         Defaults to 'SUBTRACTIVE_DITHER_1' which follows the fpack defaults
 
+    dither_seed: int or None
+        Seed for the subtractive dither.  Seeding makes the lossy compression
+        reproducible.  Allowed values are
+            None or 0 or 'clock':
+                do not set the seed explicitly, use the system clock
+            -1 or 'checksum':
```
Comment (beckermr) [inline fitsio/fitslib.py:386]:
```suggestion
            negative or 'checksum':
```

**Comment by esheldon** [comment]
OK, we'll see what Dustin says

File: fitsio/fitslib.py [hunk 6]
```diff
@@ -686,6 +699,15 @@ def write(self, data, units=None, extname=None, extver=None,
                     Preserves zeros
 
             Defaults to 'SUBTRACTIVE_DITHER_1' which follows the fpack defaults
+        dither_seed: int or None
+            Seed for the subtractive dither.  Seeding makes the lossy
+            compression reproducible.  Allowed values are
+                None or 0 or 'clock':
+                    do not set the seed explicitly, use the system clock
+                -1 or 'checksum':
```
Comment (beckermr) [inline fitsio/fitslib.py:707]:
```suggestion
                negative or 'checksum':
```

File: fitsio/fitslib.py [hunk 7]
```diff
@@ -792,6 +816,15 @@ def write_image(self, img, extname=None, extver=None,
...
```
Comment (beckermr) [inline fitsio/fitslib.py:824]:
```suggestion
                negative or 'checksum':
```

File: fitsio/fitslib.py [hunk 8]
```diff
@@ -917,7 +952,15 @@ def create_image_hdu(self,
                     Preserves zeros
 
             Defaults to 'SUBTRACTIVE_DITHER_1' which follows the fpack defaults
-
+        dither_seed: int or None
+            Seed for the subtractive dither.  Seeding makes the lossy
+            compression reproducible.  Allowed values are
+                None or 0 or 'clock':
+                    do not set the seed explicitly, use the system clock
+                -1 or 'checksum':
```
Comment (beckermr) [inline fitsio/fitslib.py:960]:
```suggestion
                negative or 'checksum':
```

File: fitsio/fitslib.py [hunk 9]
```diff
@@ -1800,6 +1845,45 @@ def get_qmethod(qmethod):
...
+            seed_out = 0
+        elif dlow == 'checksum':
+            seed_out = -1
+        else:
+            raise ValueError(f'Bad dither_seed {dither_seed}')
+    elif dither_seed is None:
+        seed_out = 0
+    else:
+        # must fit in an int
+        seed_out = numpy.int32(dither_seed).clip(min=-1)
```
Comment (beckermr) [inline fitsio/fitslib.py:1877]: Why clip here?

Comment (esheldon) [reply fitsio/fitslib.py:1877]: that's leftover from when I thought -1 was the lowest allowed

File: fitsio/fitslib.py [hunk 10]
```diff
@@ -1800,6 +1845,45 @@ def get_qmethod(qmethod):
...
+        Seed for the subtractive dither.  Seeding makes the lossy compression
+        reproducible.  Allowed values are
+            None or 0 or 'clock':
+                Return 0, do not set the seed explicitly, use the system clock
+            negative or 'checksum':
+                Return -1, means Set the seed based on the data checksum
+            1-10_000:
+                use the input seed
+    """
+    if isinstance(dither_seed, (str, bytes)):
```
[the remark this answers is not in the document]
Comment (beckermr) [reply fitsio/fitslib.py:1865]: Also this behavior is utterly baffling to me:

```
In [7]: b = b"Clock"

In [8]: b
Out[8]: b'Clock'

In [9]: type(b)
Out[9]: bytes

In [10]: b.lower()
Out[10]: b'clock'
```

If we are supposed to think of bytes as just bytes and we are forced to explicitly decode them to a string, then a method like `lower()` doesn't make sense IMHO.

**Review by beckermr — APPROVED** [review]

**Pull request merged**